src/map/map.cpp

```diff
@@ -302,34 +303,38 @@ void gamemap::overlay(const gamemap& m, const config& rules_cfg, int xpos, int y
 			const t_translation::t_terrain t = m.tiles_[x1 + m.border_size_][y1 + m.border_size_];
 			const t_translation::t_terrain current = tiles_[x2 + border_size_][y2 + border_size_];
 
-			if(t == t_translation::FOGGED || t == t_translation::VOID_TERRAIN) {
-				continue;
-			}
+			if((t != t_translation::FOGGED) && (t != t_translation::VOID_TERRAIN)) {
+				// See if there is a matching rule
+				const overlay_rule* rule = nullptr;
+				for(const overlay_rule& current_rule : rules)
+				{
+					if(!current_rule.old_.empty() && !t_translation::terrain_matches(current, current_rule.old_)) {
+						continue;
+					}
+					if(!current_rule.new_.empty() && !t_translation::terrain_matches(t, current_rule.new_)) {
+						continue;
+					}
+					rule = &current_rule;
+					break;
+				}
 
-			// See if there is a matching rule
-			const overlay_rule* rule = nullptr;
-			for(const overlay_rule& current_rule : rules)
-			{
-				if(!current_rule.old_.empty() && !t_translation::terrain_matches(current, current_rule.old_)) {
-					continue;
+				if (rule && !rule->use_old_) {
+					set_terrain(map_location(x2, y2), rule->terrain_ ? *rule->terrain_ : t , rule->mode_, rule->replace_if_failed_);
 				}
-				if(!current_rule.new_.empty() && !t_translation::terrain_matches(t, current_rule.new_)) {
-					continue;
+				else {
+					set_terrain(map_location(x2,y2), t);
 				}
-				rule = &current_rule;
-				break;
 			}
 
-			if (rule && !rule->use_old_) {
-				set_terrain(map_location(x2, y2), rule->terrain_ ? *rule->terrain_ : t , rule->mode_, rule->replace_if_failed_);
-			}
-			else {
-				set_terrain(map_location(x2,y2), t);
-				if (false /*TODO: add overwrite special locs paramter*/) {
-					starting_positions_.right.erase(t_translation::coordinate(x2, y2));
-					for (auto& pair : m.starting_positions_.right.equal_range(t_translation::coordinate(x1, y1))) {
-						starting_positions_.insert(tstarting_positions::value_type(pair.second, t_translation::coordinate(x2, y2)));
-					}
+			// TODO (GL 12SEP2016) This does not check to ensure a location given on the
+			// terrain mask does not exist elsewhere on the underlying map, in an area not
+			// effected by the mask. That requires a name look-up against the map and removing
+			// the starting postition, if found. At present this simply removes by location as
+			// we work through the overlay.
+			if (overlay_special_locations) {
+				starting_positions_.right.erase(t_translation::coordinate(x2, y2));
+				for (auto& pair : m.starting_positions_.right.equal_range(t_translation::coordinate(x1, y1))) {
+					starting_positions_.insert(tstarting_positions::value_type(pair.second, t_translation::coordinate(x2, y2)));
```

Comment:
Note that starting_positions_ is a boost::bimaps::bimap so it might not be as easy as starting_positions_[key].

I agree that overwriting would be better here.